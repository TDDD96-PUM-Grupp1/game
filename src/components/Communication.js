import createDeepstream from 'deepstream.io-client-js';

class Communication {
  /*
  * Constructor for Communication.
  * Establishes connection with the Deepstream server. Awaits
  * a RPC from a client.
  */
  constructor() {
    // Connect to deepstream
    this.ds = createDeepstream('10.90.128.65:60020');
    this.instance = 'abc'; // this.client.getUid();
    this.client = this.ds.login({ username: this.instance });
    this.players = {};
    this.addPlayer = this.addPlayer.bind(this);
    this.client.rpc.provide(`data/${this.instance}/addPlayer`, this.addPlayer);

    // Bind callbacks
    this.getPlayers = this.getPlayers.bind(this);
    this.readSensorData = this.readSensorData.bind(this);
    this.presenceUpdate = this.presenceUpdate.bind(this);

  }

  /*
  * Adds player to the list of players. Starts to listen to the events
  * published by that player.
  */
  addPlayer(data, response) {
    console.log(`Player ${data.id} has connected`);
    this.players[data.id] = { name: data.name, sensor: data.sensor };
    this.client.event.subscribe(`data/${this.instance}/${data.id}`, this.readSensorData);
    response.send(data.id);
    // this.client.presence.subscribe(data.id, this.presenceUpdate);
  }

  /*
  * Callback for a presence update for a controller.If disconnected
  * removes player and unsubscribes from player channel.
  */
  presenceUpdate(username, login) {
    if (!login) {
      console.log(`Players ${username} has disconnected`);
      delete this.players[username];
      this.client.event.unsubscribe(`data/${this.instance}/${username}`);
      this.client.presence.unsubscribe(username);
    }
  }

  /*
  * Callback for reading sensordata. Updates a specific players sensorvalues.
  */
  readSensorData(data) {
    if (data.sensor) {
      this.players[data.id].sensor = data.sensor;
    }
  }

  getPlayers() {
    return this.players;
  }
}

export default Communication;
