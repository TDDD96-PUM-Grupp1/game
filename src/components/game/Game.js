// import * as PIXI from 'pixi.js';
import EntityHandler from './EntityHandler';
import TestGamemode from './gamemodes/TestGamemode';

/* eslint-disable */
var GAME;
/* eslint-enable */

/*
Game.
*/
class Game {
  constructor(app) {
    this.app = app;

    // Set global variables for the pixi context and game context to heavily simplify
    // the creation of additional objects and graphics when necessary.
    GAME = this;

    // Create all handlers
    this.entityHandler = new EntityHandler();

    // Create gamemode
    this.currentGamemode = new TestGamemode(this);
  }

  // Main game loop
  loop(delta) {
    // Convert frame delta to time delta [second] (assuming 60fps)
    const dt = delta / 60;

    // Update object handler
    this.currentGamemode.preUpdate(dt);
    this.entityHandler.update(dt);
    this.currentGamemode.postUpdate(dt);
  }
}

export default Game;
