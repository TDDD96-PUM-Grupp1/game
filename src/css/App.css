.App {
  text-align: center;
}

.game-title{
  font-size: 70px;
  font-family: 'Amatic SC';

  margin-bottom: 40px;
}

.center-menu{
  margin-top: 50px;
}

.menu-button{
  display:block;
  margin: auto;

  margin-bottom: 20px;
  height: 60px;
  width: 250px;

  color: #FFFFFF;
  background-color: #F49F45;
  font-size: 20px;

  cursor:pointer;

  border-radius: 10px;
  border: none;
  border-bottom: 5px #D9842A solid;
}

.menu-button:hover{
  background-color: #FF8300;
}

.menu-button-holder{
  background-color: #ffecbd;

  border-radius: 30px;
  display: inline-block;

  padding: 30px;

  margin:auto;
}
