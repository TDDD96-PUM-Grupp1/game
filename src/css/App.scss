@import 'globals';

@include react-md-everything;
.App {
  text-align: center;
}
