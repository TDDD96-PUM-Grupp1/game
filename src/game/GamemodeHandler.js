import settings from '../config';
import GamemodeConfig from './GamemodeConfig';

import TestGamemode from './gamemodes/TestGamemode';
import KnockOff from './gamemodes/KnockOff';
import KnockOffRandom from './gamemodes/KnockOffRandom';
import KnockOffDynamic from './gamemodes/KnockOffDynamic';
import KnockOffWander from './gamemodes/KnockOffWander';
import KnockOffSpinner from './gamemodes/KnockOffSpinner';
import DodgebotBumper from './gamemodes/DodgebotBumper';
import Dodgebot from './gamemodes/Dodgebot';
import Hockey from './gamemodes/Hockey';
import PassTheBomb from './gamemodes/PassTheBomb';

/**
 * Singleton class for handling gamemode storage and selection
 */
class GMHandlerClass {
  constructor() {
    // Map of all available gamemodes
    this.gamemodes = {};
    // Map of the gamemodes configs
    this.configs = {};
    // Map of gamemode rules
    this.rules = {};

    this.addGamemodes();

    this.selected = '';
    if (settings.skipmenu) {
      this.selectGameMode(settings.defaultGamemode);
    }
  }

  addGamemodes() {
    this.addGamemode('Knock Off', KnockOff);
    this.addGamemode('Dodgebot', Dodgebot);
    this.addGamemode('Knock Off Random', KnockOffRandom, false);
    this.addGamemode('Knock Off Dynamic', KnockOffDynamic, false);
    this.addGamemode('Knock Off Wander', KnockOffWander);
    this.addGamemode('Test Gamemode', TestGamemode, false);
    this.addGamemode('Knock Off Spinner', KnockOffSpinner);
    this.addGamemode('Dodgebot Bumpers', DodgebotBumper);
    this.addGamemode('Hockey', Hockey);
    this.addGamemode('Pass The Bomb', PassTheBomb);
  }

  addGamemode(name, Gamemode, visible = true) {
    const options = Gamemode.getConfig();
    const resources = Gamemode.getResources();

    this.rules[name] = options.rules;
    this.gamemodes[name] = { Gamemode, visible };
    this.configs[name] = new GamemodeConfig(name, resources, options);
  }

  /**
   Get list of all names of available gamemodes
   */
  getGamemodes() {
    const list = [];
    // Only add visible gamemodes to the list
    Object.keys(this.gamemodes).forEach(name => {
      const { visible } = this.gamemodes[name];
      if (visible) {
        list.push(name);
      }
    });
    return list;
  }

  /**
   Set the selected gamemode
   */
  selectGameMode(name) {
    if (name in this.gamemodes) {
      this.selected = name;
    } else {
      throw new Error(`Gamemode ${name} does not exist`);
    }
  }

  /**
   Get which gamemode has been selected
   */
  getSelected() {
    if (!this.selected) {
      throw new Error('Gamemode has not been selected');
    }

    const { Gamemode } = this.gamemodes[this.selected];
    const { resources, options } = this.configs[this.selected];

    return {
      SelectedMode: Gamemode,
      requestedResources: resources,
      options,
    };
  }

  /**
   Get description of gamemode rules.
   */
  getRules(gamemode) {
    return this.rules[gamemode];
  }

  /**
   * Get the selected mode identifier, this will return the name
   * of the current gamemode.
   */
  getSelectedId() {
    return this.selected;
  }

  /**
   * Returns an array containing button information for the selected game-mode
   */
  getButtons() {
    const { options } = this.configs[this.selected];

    const buttons = [];
    if (options.abilities) {
      options.abilities.forEach(ability => {
        buttons.push({
          name: ability.name,
          cooldown: ability.cooldown,
          duration: ability.duration,
          color: ability.color,
        });
      });
    }
    return buttons;
  }
}
/**
 * Singleton Pattern from http://www.dofactory.com/javascript/singleton-design-pattern
 */
const GamemodeHandler = (() => {
  let instance;

  const createInstance = () => new GMHandlerClass();

  return {
    getInstance: () => {
      if (!instance) {
        instance = createInstance();
      }

      return instance;
    },
  };
})();

export default GamemodeHandler;
