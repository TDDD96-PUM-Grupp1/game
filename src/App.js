import React from 'react';
import logo from './logo.svg';
import './App.css';
import Communication from './components/Communication';
import PlayerList from './components/PlayerList';
import StartMenu from './components/StartMenu';

const com = new Communication();

const App = () => (
  <div className="App">
    <header className="App-header">
      <img src={logo} className="App-logo" alt="logo" />
      <h1 className="App-title">Welcome to React</h1>
    </header>
    <p className="App-intro">
      To get started, edit <code>src/App.js</code> and save to reload.
    </p>
    <StartMenu />
    <PlayerList getPlayersFunction={com.getPlayers} />

  </div>
);

export default App;
