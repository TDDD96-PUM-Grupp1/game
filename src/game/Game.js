// import * as PIXI from 'pixi.js';
import EntityHandler from './EntityHandler';
import CollisionHandler from './CollisionHandler';
import ResourceServer from './ResourceServer';
import GamemodeHandler from './GamemodeHandler';
import RespawnHandler from './RespawnHandler';

/*
Game.
*/
class Game {
  constructor(app, communication) {
    this.app = app;
    this.communication = communication;
    this.instance = this.communication.getInstance();

    // This will be undefined when running tests since we havn't
    // started an instance.
    if (this.instance !== undefined) {
      this.instance.addInstanceListener(this);
    }

    // Create all handlers
    this.entityHandler = new EntityHandler();
    this.collisionHandler = new CollisionHandler(this.entityHandler);
    this.respawnHandler = new RespawnHandler(this.entityHandler);
    this.resourceServer = new ResourceServer();

    // Create gamemode
    const gamemodeHandler = GamemodeHandler.getInstance();
    const SelectedMode = gamemodeHandler.getSelected();
    this.currentGamemode = new SelectedMode(this);
  }

  // Main game loop
  loop(delta) {
    // Convert frame delta to time delta [second] (assuming 60fps)
    const dt = delta / 60;

    // Update handlers and gamemodes
    this.currentGamemode.preUpdate(dt);
    this.entityHandler.update(dt);
    this.collisionHandler.handleCollisions(dt);
    this.currentGamemode.postUpdate(dt);
    this.respawnHandler.checkRespawns();
    this.entityHandler.updateGraphics(dt);

    this.communication.update(dt);
  }

  // Called when a new player joins.
  onPlayerJoin(idTag) {
    this.currentGamemode.onPlayerJoin(idTag);
  }

  // Called when a player leaves the game.
  onPlayerLeave(idTag) {
    this.currentGamemode.onPlayerLeave(idTag);
  }
  // eslint-disable-next-line
  onSensorMoved(id, sensor) {}

  onButtonsPressed(id, button) {
    console.log('Button pressed ')
    console.log(button);
  }

  // eslint-disable-next-line
  onButtonPressed(id, button) {}
}

export default Game;
