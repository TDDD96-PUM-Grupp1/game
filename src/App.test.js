import React from 'react';
import ReactDOM from 'react-dom';
import assert from 'assert';
import * as PIXI from 'pixi.js';

import settings from './config';
import Communication from './components/Communication';
import App from './App';
import GamemodeHandler from './game/GamemodeHandler';
import Game from './game/Game';

it('renders without crashing', () => {
  const div = document.createElement('div');
  ReactDOM.render(<App />, div);
  ReactDOM.unmountComponentAtNode(div);
});

describe('GamemodeHandler', () => {
  it('is a working singleton', () => {
    const inst1 = GamemodeHandler.getInstance();
    const inst2 = GamemodeHandler.getInstance();

    assert.equal(inst1, inst2);
  });

  it('selects all gamemodes', () => {
    const gmHandler = GamemodeHandler.getInstance();
    const gmList = gmHandler.getGamemodes();

    for (let i = 0; i < gmList.length; i += 1) {
      gmHandler.selectGameMode(gmList[i]);

      // Check so selection is done
      gmHandler.getSelected();
    }
  });

  it('loads all gamemodes', () => {
    const gmHandler = GamemodeHandler.getInstance();
    const gmList = gmHandler.getGamemodes();
    let GamemodeClass;
    // eslint-disable-next-line no-unused-vars
    let gamemode;

    // Extra components to allow for loading gamemodes
    const pixi = new PIXI.Application();
    const com = new Communication(settings.communication, () => {});
    const game = new Game(pixi, com);

    for (let i = 0; i < gmList.length; i += 1) {
      gmHandler.selectGameMode(gmList[i]);

      GamemodeClass = gmHandler.getSelected();
      gamemode = new GamemodeClass(game);
    }
  });
});
