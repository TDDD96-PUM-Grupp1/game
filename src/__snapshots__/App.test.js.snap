// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FirstMenu matches the snapshot 1`] = `
<div>
  <button
    className="md-btn md-btn--raised md-btn--text md-pointer--hover md-background--primary md-background--primary-hover md-inline-block menu-button"
    disabled={undefined}
    href={undefined}
    onClick={[Function]}
    onKeyDown={[Function]}
    onKeyUp={[Function]}
    onMouseDown={[Function]}
    onMouseEnter={[Function]}
    onMouseLeave={[Function]}
    onMouseUp={[Function]}
    onTouchEnd={[Function]}
    onTouchStart={[Function]}
    type="button"
  >
    <div
      className="md-ink-container"
      style={undefined}
    />
    Create Game
  </button>
  <button
    className="md-btn md-btn--raised md-btn--text md-pointer--hover md-background--primary md-background--primary-hover md-inline-block menu-button"
    disabled={undefined}
    href={undefined}
    onClick={[Function]}
    onKeyDown={[Function]}
    onKeyUp={[Function]}
    onMouseDown={[Function]}
    onMouseEnter={[Function]}
    onMouseLeave={[Function]}
    onMouseUp={[Function]}
    onTouchEnd={[Function]}
    onTouchStart={[Function]}
    type="button"
  >
    <div
      className="md-ink-container"
      style={undefined}
    />
    About
  </button>
</div>
`;
